### PR TITLE
Fix n-m db access in ProjectResource.getProjects to add projectsWithLatestRevision

### PR DESCRIPTION
**Pull request by katsuyan**
## Overview
fix https://github.com/treasure-data/digdag/blob/eb44c91fbee3bc38f869c088814a2a623b831b3d/digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java#L263

I fix n-m db access in `ProjectResource.getProjects` to add `projectWithLatestRevision`.

## Reason for fix
I try to change SQL `order by` value dynamically in https://github.com/treasure-data/digdag/pull/1156.
But, getProjects and getRevision SQL separated. So, I can't sort prejects with revision by any value in SQL.

### CODE REVIEW

**Comment by yoyama** [comment]
@katsuyan Thank you for your PR.
I checked it and basically looks good.
But I have one concern about performance of the API when project and revision increase.
Because the new DB method use multiple join in the SQL.
Did you confirm around it?
If so please let me know, how to test and the result.

**Comment by yoyama** [comment]
@muga If you have a chance, please check this PR.

**Comment by katsuyan** [comment]
@yoyama  Thank you your comment!
I test query performance.

### Environment
- OS: macOS 10.13.6
- CPU: 3.1 GHz Intel Core i7
- MEM: 16 GB 2133 MHz LPDDR3

### Create dummy data
I create dummy data in this query .

```
insert into projects (id, site_id, name, created_at)
select
  i,
  0,
  format('p%s', i),
  clock_timestamp()
from
  generate_series(0, 1000000) as i
;


insert into revisions (project_id, name, archive_type, archive_md5, created_at)
select
  i % 1000000,
  format('1c2e7c9f-aaba-49b1-9544-e0686dd%s', i),
  'db',
  '\x2c95301a1883329e89a1c352bddc9d11',
  clock_timestamp()
from
  generate_series(0, 3000000) as i
;
```

### Measure time
I test query performance. Sorry, This query is so slow.  I find out `order by proj.id asc` is bottleneck.

```
select proj.*, rev.name as revision_name, rev.created_at as revision_created_at, rev.archive_type as revision_archive_type, rev.archive_md5 as revision_archive_md5
from projects proj
join revisions rev on proj.id = rev.project_id
join (
  select project_id, max(id) AS id
  from revisions
  group by project_id
) a on a.id = rev.id
where proj.site_id = 0
and proj.name is not null
and proj.id > 0
order by proj.id asc
limit 100;

time: 6658.709 ms
```

### Tuning

So I tuned the query. 

```
select * from
(
  select proj.*, rev.name as revision_name, rev.created_at as revision_created_at, rev.archive_type as revision_archive_type, rev.archive_md5 as revision_archive_md5, rev.id as revision_id, max(rev.id) OVER(partition by rev.project_id) as max_revision_id
  from projects proj
  join revisions rev on proj.id = rev.project_id
  where proj.site_id = 0
  and proj.name is not null
  and proj.id > 0
  order by proj.id asc
) as projects_with_revision
where projects_with_revision.revision_id = projects_with_revision.max_revision_id
limit 100;

time: 2.965 ms
```

This query only be used in PostgreSQL. I could't tuned query in H2.
So I try to use this query only in PostgreSQL and I will give up tune in H2.
What do you think?


### Appendix
I try to use below query. But, these queries slow too.

```
select proj.*, rev.name as revision_name, rev.created_at as revision_created_at, rev.archive_type as revision_archive_type, rev.archive_md5 as revision_archive_md5
from projects proj
join revisions rev on proj.id = rev.project_id
where rev.id = any(array(
      select max(id)
      from revisions
      group by project_id
))
and proj.site_id = 0
and proj.name is not null
and proj.id > 0
order by proj.id asc
limit 100;

tiem: 4054.134 ms
```

```
select proj.*, rev.name as revision_name, rev.created_at as revision_created_at, rev.archive_type as revision_archive_type, rev.archive_md5 as revision_archive_md5
from
(select * from projects
where site_id = 0
and name is not null
and id > 0
order by id asc
limit 100) as proj
join revisions rev on proj.id = rev.project_id
join (
  select project_id, max(id) AS id
  from revisions
  group by project_id
) a on a.id = rev.id;

time: 1419.528 ms
```

**Comment by yoyama** [comment]
@katsuyan Thanks for your test. It is good try!
I have a question on your fixed query.
```
select * from
( select ...
  order by proj.id asc
) as projects_with_revision
where projects_with_revision.revision_id = projects_with_revision.max_revision_id
limit 100;
```
In theses query, you don't specify _order by_ in outermost SQL.
In this case, Does PostgreSQL really ensure the order of sub-qeury?

**Comment by katsuyan** [comment]
I change `order by` location out side subquery.
And I checked it is same execution plan.


```
digdag=# explain select * from
digdag-# (
digdag(#   select proj.*, rev.name as revision_name, rev.created_at as revision_created_at, rev.archive_type as revision_archive_type, rev.archive_md5 as revision_archive_md5, rev.id as revision_id, max(rev.id) OVER(partition by rev.project_id) as max_revision_id
digdag(#   from projects proj
digdag(#   join revisions rev on proj.id = rev.project_id
digdag(#   where proj.site_id = 0
digdag(#   and proj.name is not null
digdag(#   and proj.id > 0
digdag(# ) as projects_with_revision
digdag-# where projects_with_revision.revision_id = projects_with_revision.max_revision_id
digdag-# order by id asc
digdag-# limit 100;
                                                           QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=30.90..2670.21 rows=100 width=137)
   ->  Subquery Scan on projects_with_revision  (cost=30.90..395928.05 rows=15000 width=137)
         Filter: (projects_with_revision.revision_id = projects_with_revision.max_revision_id)
         ->  WindowAgg  (cost=30.90..358427.92 rows=3000010 width=141)
               ->  Merge Join  (cost=30.90..313427.77 rows=3000010 width=137)
                     Merge Cond: (proj.id = rev.project_id)
                     ->  Index Scan using projects_pkey on projects proj  (cost=0.42..37353.43 rows=1000000 width=63)
                           Index Cond: (id > 0)
                           Filter: ((name IS NOT NULL) AND (site_id = 0))
                     ->  Index Scan using revisions_on_project_id on revisions rev  (cost=0.43..236076.45 rows=3000013 width=74)
```

```
digdag=# explain select * from
digdag-# (
digdag(#   select proj.*, rev.name as revision_name, rev.created_at as revision_created_at, rev.archive_type as revision_archive_type, rev.archive_md5 as revision_archive_md5, rev.id as revision_id, max(rev.id) OVER(partition by rev.project_id) as max_revision_id
digdag(#   from projects proj
digdag(#   join revisions rev on proj.id = rev.project_id
digdag(#   where proj.site_id = 0
digdag(#   and proj.name is not null
digdag(#   and proj.id > 0
digdag(#   order by proj.id asc
digdag(# ) as projects_with_revision
digdag-# where projects_with_revision.revision_id = projects_with_revision.max_revision_id
digdag-# limit 100;
                                                           QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=30.90..2670.21 rows=100 width=137)
   ->  Subquery Scan on projects_with_revision  (cost=30.90..395928.05 rows=15000 width=137)
         Filter: (projects_with_revision.revision_id = projects_with_revision.max_revision_id)
         ->  WindowAgg  (cost=30.90..358427.92 rows=3000010 width=141)
               ->  Merge Join  (cost=30.90..313427.77 rows=3000010 width=137)
                     Merge Cond: (proj.id = rev.project_id)
                     ->  Index Scan using projects_pkey on projects proj  (cost=0.42..37353.43 rows=1000000 width=63)
                           Index Cond: (id > 0)
                           Filter: ((name IS NOT NULL) AND (site_id = 0))
                     ->  Index Scan using revisions_on_project_id on revisions rev  (cost=0.43..236076.45 rows=3000013 width=74)
```

So, I change optimized query that `order by` location is out side subquery.


## Performance Test
I compare time `digdag projects` between master and this branch.

### Preparation
When I mesure `digdag projects` time, I fix below.

https://github.com/treasure-data/digdag/blob/master/digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java#L264

to

```
collection = ps.getProjects(10000, Optional.absent())
```


### Result
master
```
Use `digdag workflows <project-name>` to show details.
       11.22 real         5.83 user         0.43 sys
```

this branch(PostgreSQL)
```
Use `digdag workflows <project-name>` to show details.
        2.93 real         6.12 user         0.40 sys
```

**Comment by yoyama** [comment]
@katsuyan  Thanks for your modification. I will check it later.
Please wait for a while.

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseProjectStoreManager.java [hunk 1]
```diff
@@ -508,6 +516,24 @@ public void deleteSchedules(int projId)
     public interface H2Dao
             extends Dao
     {
+        @Override
+        // h2 doesn't have max window function.
+        // this is not efficient but gave up optimization on h2.
```
Comment (yoyama) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseProjectStoreManager.java]: @katsuyan Trivial thing, I hope this comment move upon the Override annotation.

**Comment by yoyama** [comment]
@katsuyan  Thank you for your contribution. I commented on one very small thing. 
Could you check it?.

I am testing your branch and it looks like no problem. So after your fix, I will approve this PR.

**Comment by katsuyan** [comment]
Thank you your check!
I fixed comment location.

**Comment by yoyama** [comment]
@katsuyan  LGTM. Thank you for your contribution!

**Pull request merged**